# StrainPanDA - A strain analysis pipeline based on pangenome

StrainPanDA is a tool that deconvolutes pangenome coverage into strain composition and strain gene profile. We provide a fully automated pipeline implemented with [nextflow](https://www.nextflow.io/docs/latest/index.html).

## Installation

1. install nextflow

```sh
curl -s https://get.nextflow.io | bash
```

2. We provide two dockers required by the nextflow pipeline (Step 2 and Step 3). Recommended!

Step2:PanPhlAn (for mapping short reads to pan-genome databases)

```
docker pull yuxiangtan/strainpanda-mapping:dev
docker tag yuxiangtan/strainpanda-mapping:dev strainpanda-mapping:dev
```
Step3. Decomposing pangenome into strains in R
```
docker pull yuxiangtan/strainpanda-strainpandar:dev
docker tag yuxiangtan/strainpanda-strainpandar:dev strainpanda-strainpandar:dev
```

3. check the dockers (First, the dockers should be pulled completely)
```
docker run -u $(id -u):$(id -g) strainpanda-mapping:dev panphlan_profile.py -h
docker run -u $(id -u):$(id -g) strainpanda-strainpandar:dev R --no-save
```


### Local installation, only use it if docker is not available
Step2. install PanPhlAn (for mapping short reads to pan-genome databases)

https://github.com/SegataLab/panphlan/wiki/Home-1_3


Step3. install [R](https://www.r-project.org/) and [strainpandar](src/strainpandar): decomposing pangenome into strains

Install required packages `dplyr`, `foreach`, `MASS`, `NMF`, `pracma`, `vegan`

```sh
git clone https://github.com/xbiome/StrainPanDA.git
cd StrainPanDA/src
tar -czf strainpandar.tar.gz strainpandar
R CMD INSTALL strainpandar.tar.gz
```



## DataBase Preparation
### Pre-built pangenome databases

Database can be found from Zenodo (doi:10.5281/zenodo.6592017): [StrainPanDA Pre-built pangenome database | Zenodo](https://zenodo.org/record/6592017)

In each tar.gz file:

 - The pangenome file to be used is the `${species_version}_pangenome.csv` file, and the pangenome.csv is the recongizer by StrainPanDA;
 - The pangenome sequences are in the `${species_version}_centroids.ffn` file;
 - The bowtie2 indexes are the `bt2` files;
 - The pangenome of the species is annotated by the `${species_version}_emapper_anno.tsv` file, with the colunms specified in the header.
 - The vfdb (Apr 9th 2021) annotation is the `${species_version}_vfdb_anno.csv` file, with the first column as the gene family ID and the second column as the VFDB ID.
 - The CAZy (Jul 31st 2019) annotation is the `${species_version}_CAZy_anno.csv` file, with the first column as the gene family ID and the second column as the CAZy catalog ID.

<br>
Each tar.gz file must be unzipped before using:

```
#within the path of tar.gz files
tar -zxvf Acinetobacter-johnsonii-202009.tar.gz #unzip only the database of Acinetobacter-johnsonii
tar -zxvf *.tar.gz #unzip all the database within the folder
```

### User-specific database generation

User-specific database with annotation could be generated by [the following way](https://github.com/xbiome/StrainPanDA/tree/main/custom_db#readme). 

## Run analysis

### Run full analysis
Download test data to a folder e.g. `reads`

```sh
wget ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR581/005/SRR5813295/SRR5813295_1.fastq.gz
wget ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR581/005/SRR5813295/SRR5813295_2.fastq.gz
wget ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR581/006/SRR5813296/SRR5813296_1.fastq.gz
wget ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR581/006/SRR5813296/SRR5813296_2.fastq.gz
wget ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR581/007/SRR5813297/SRR5813297_1.fastq.gz
wget ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR581/007/SRR5813297/SRR5813297_2.fastq.gz
wget ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR581/008/SRR5813298/SRR5813298_1.fastq.gz
wget ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR581/008/SRR5813298/SRR5813298_2.fastq.gz
wget ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR581/009/SRR5813299/SRR5813299_1.fastq.gz
wget ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR581/009/SRR5813299/SRR5813299_2.fastq.gz
```

Create a species list

```sh
echo "Faecalibacterium prausnitzii" > species_list.txt
```

#### With docker

```sh
/home/bin/StrainPanDA/main.nf -profile docker --ref_path /home/database/StrainPanDA --path reads/ --ref_list species_list.txt
```
Here, /home/bin/StrainPanDA is the PATH_TO_REPO; /home/database/StrainPanDA is the PATH_TO_REFERENCE; reads/ is the PATH_TO_RAW_FASTQ


#### Local

```sh
PATH_TO_REPO/main.nf --ref_path PATH_TO_REFERENCE --path reads/ --ref_list species_list.txt
```

### Run strainpandar

Assuming you have a count matrix generated, we provide a [Rscript wrapper](bin/run_strainpandar.r) to run strainpandar. A sample count matrix can be found [here](data/Faecalibacterium-prausnitzii-202009.counts.csv).


```sh
Rscript bin/run_strainpandar.r -c data/Faecalibacterium-prausnitzii-202009.counts.csv -r data/refs/Faecalibacterium-prausnitzii-202009 -o work -t 8 -m 8 -n 0
```

The parameters of the wrapper can be found from the help message:

```sh
> Rscript bin/run_strainpandar.r -h
A wrapper script to perform strain decomposition using strainpandar package.
Usage: bin/run_strainpandar.r [-[-help|h]] [-[-counts|c] <character>] [-[-reference|r] <character>] [-[-output|o] [<character>]] [-[-threads|t] [<integer>]] [-[-max_rank|m] [<integer>]] [-[-rank|n] [<integer>]]
    -h|--help         Show this help message
    -c|--counts       Gene-sample count matrix (CSV file) obtained from mapping reads to a reference pangenome [required]
    -r|--reference    Pangenome database path [required]
    -o|--output       Output prefix [default: ./strainpandar]
    -t|--threads      Number of threads to run in parallele [default: 1]
    -m|--max_rank     Max number of strains expected [default: 8]
    -n|--rank         Number of strains expected. If 0, try to select from 1 to `max_rank`. If not 0, overwrite `max_rank`. [default: 0]
```


## Outputs

Main outputs of the pipeline:

 - Pangenome database mapping outputs
   - Merged count matrix `{species-version}.counts.csv`: each row is one gene family, each column is one sample, values are read counts.
 - Strain decomposition outputs
   - Gene family-strain matrix (**P**) `{species-version}.strainpanda.genefamily_strain.csv`: each row is one gene family, each column is one strain, values (binary) are presence (1) or absence (0) of the gene family.
   - Strain-sample matrix (**S**) `{species-version}.strainpanda.strain_sample.csv`: each row is one strain, each column is one sample, values are the relative abundances of the strain (in fraction).

Ouput files from the above run:

```sh
strainpanda_out/
├── Faecalibacterium-prausnitzii-202009.counts.csv  ## Merged count matrix (gene family by sample)
├── Faecalibacterium-prausnitzii-202009_mapping     ## Sample specific count files
│   ├── SRR5813295_Faecalibacterium-prausnitzii-202009.csv.bz2
│   ├── SRR5813296_Faecalibacterium-prausnitzii-202009.csv.bz2
│   ├── SRR5813297_Faecalibacterium-prausnitzii-202009.csv.bz2
│   ├── SRR5813298_Faecalibacterium-prausnitzii-202009.csv.bz2
│   └── SRR5813299_Faecalibacterium-prausnitzii-202009.csv.bz2
├── Faecalibacterium-prausnitzii-202009_strainpandar_out ## strainpandar outputs
│   ├── Faecalibacterium-prausnitzii-202009.strainpanda.anno_strain_sample.pdf ## annotation to the closest reference
│   ├── Faecalibacterium-prausnitzii-202009.strainpanda.genefamily_strain.csv ## gene family-strain matrix
│   ├── Faecalibacterium-prausnitzii-202009.strainpanda.genefamily_strain.pdf ## heatmap visualization
│   ├── Faecalibacterium-prausnitzii-202009.strainpanda.rds ## R object contains strainpandar results
│   ├── Faecalibacterium-prausnitzii-202009.strainpanda.strain_sample.csv ## strain-sample matrix
│   └── Faecalibacterium-prausnitzii-202009.strainpanda.strain_sample.pdf ## barplot visualization
└── pipeline_info  ## pipeline run statistics
    ├── strainpanda_DAG.svg
    ├── strainpanda_report.html
    ├── strainpanda_timeline.html
    └── strainpanda_trace.txt
```



## Examples

All the examples used in the manuscript can be viewed [here](https://github.com/xbiome/StrainPanDA-data/tree/main/example#readme)

.
